File: lib/src/main/java/io/github/raphiz/zackbumm/FileSystemWatcher.java
package io.github.raphiz.zackbumm;

import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.Set;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.logging.Logger;

class FileSystemWatcher {
    private final Logger logger = LoggerHelpers.logger();
    private final AtomicBoolean active = new AtomicBoolean(false);
    private final Set<Path> watchDirectories;
    private final Consumer<FileSystemEvent> onChange;
    private final WatchService watchService;

    public FileSystemWatcher(Set<Path> watchDirectories, Consumer<FileSystemEvent> onChange) throws IOException {
        this.watchDirectories = watchDirectories;
        this.onChange = onChange;
        this.watchService = FileSystems.getDefault().newWatchService();
    }

    private void watchRecursively(Path path, boolean notify) throws IOException {
        // TODO: Handle creation/deletion of watchDirectories
        if(!Files.exists(path)) return;
        Files.walkFileTree(path, new SimpleFileVisitor<>() {
            @Override
            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
                if (notify) {
                    onChange.accept(new FileSystemEvent(file.toAbsolutePath(), EventType.CREATED));
                }
                return FileVisitResult.CONTINUE;
            }

            @Override
            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
                if (!Files.isHidden(dir)) {
                    dir.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE);
                    return FileVisitResult.CONTINUE;
                } else {
                    return FileVisitResult.SKIP_SUBTREE;
                }
            }
        });
    }

    public void start() throws IOException {
        logger.info("Starting watch service");
        for (Path watchDirectory : watchDirectories) {
            logger.fine("Watching directory " + watchDirectory);
            watchRecursively(watchDirectory, false);
        }
        active.set(true);
        Executors.newSingleThreadExecutor().submit(() -> {
            Thread.currentThread().setName("filewatch-thread");
            while (active.get()) {
                WatchKey watchKey;
                try {
                    watchKey = watchService.poll(300, TimeUnit.MILLISECONDS);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
                if (watchKey == null) {
                    continue;
                }
                Path directory = (Path) watchKey.watchable();
                for (WatchEvent<?> event : watchKey.pollEvents()) {
                    if (event.kind() == StandardWatchEventKinds.OVERFLOW) {
                        logger.warning("WatchService Overflow occurred");
                    } else {
                        Path changedFile = directory.resolve((Path) event.context());
                        if (Files.isDirectory(changedFile)) {
                            try {
                                watchRecursively(changedFile, true);
                            } catch (IOException e) {
                                logger.warning("Failed to watch directory " + changedFile + ": " + e.getMessage());
                            }
                        } else {
                            logger.fine(() -> "Received event " + event.kind().name() + " for file " + changedFile);
                            onChange.accept(new FileSystemEvent(changedFile.toAbsolutePath(), toEventType(event)));
                        }
                    }
                }

                boolean keyIsValid = watchKey.reset();
                if (!keyIsValid) {
                    logger.fine(() -> directory + " has been unregistered");
                    if (watchDirectories.contains(directory)) {
                        logger.info("Watch directory " + directory + " has been unregistered. Will stop watch service");
                        try {
                            stop();
                        } catch (IOException e) {
                            throw new RuntimeException(e);
                        }
                    }
                }
            }
        });
    }

    public void stop() throws IOException {
        logger.info("Stopping watch service");
        active.set(false);
        watchService.close();
    }

    private EventType toEventType(WatchEvent<?> event) {
        WatchEvent.Kind<?> kind = event.kind();
        if (kind == StandardWatchEventKinds.ENTRY_MODIFY) {
            return EventType.MODIFIED;
        } else if (kind == StandardWatchEventKinds.ENTRY_DELETE) {
            return EventType.DELETED;
        } else if (kind == StandardWatchEventKinds.ENTRY_CREATE) {
            return EventType.CREATED;
        } else {
            throw new IllegalStateException("Unknown event kind " + kind + " received");
        }
    }
}

record FileSystemEvent(Path path, EventType eventType) {
}

File: lib/src/main/kotlin/io/github/raphiz/zackbum/FileSystemWatcher.kt
package io.github.raphiz.zackbum

import io.github.raphiz.zackbum.LoggerHelpers.logger
import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.io.path.isDirectory
import kotlin.io.path.isHidden

internal class FileSystemWatcher(
    private val workspace: Path,
    private val onChange: (FileSystemEvent) -> Unit,
) {
    private val logger = logger()
    private val active = AtomicBoolean(false)

    private val watchService = FileSystems.getDefault().newWatchService()

    private fun Path.watchRecursively(notify: Boolean = false) {
        Files.walkFileTree(
            this,
            object : SimpleFileVisitor<Path>() {
                override fun visitFile(
                    file: Path,
                    attrs: BasicFileAttributes,
                ): FileVisitResult {
                    if (notify) {
                        onChange(FileSystemEvent(file, EventType.CREATED))
                    }
                    return FileVisitResult.CONTINUE
                }

                override fun preVisitDirectory(
                    dir: Path,
                    attrs: BasicFileAttributes,
                ): FileVisitResult =
                    if (dir.isDirectory() && !dir.isHidden()) {
                        dir.register(
                            watchService,
                            StandardWatchEventKinds.ENTRY_MODIFY,
                            StandardWatchEventKinds.ENTRY_CREATE,
                            StandardWatchEventKinds.ENTRY_DELETE,
                        )
                        FileVisitResult.CONTINUE
                    } else {
                        FileVisitResult.SKIP_SUBTREE
                    }
            },
        )
    }

    fun start() {
        logger.fine { "Starting watch service on workspace ${workspace.toAbsolutePath()}" }
        workspace.watchRecursively()
        active.set(true)
        Executors.newSingleThreadExecutor().submit {
            Thread.currentThread().setName("filewatch-thread")
            while (active.get()) {
                val watchKey = watchService.poll(300, TimeUnit.MILLISECONDS) ?: continue
                val directory = watchKey.watchable() as Path
                watchKey.pollEvents().forEach { event ->
                    if (event.kind() == StandardWatchEventKinds.OVERFLOW) {
                        logger.warning("WatchService Overflow occurred")
                    } else {
                        val changedFile = directory.resolve(event.context() as Path)
                        if (changedFile.isDirectory()) {
                            changedFile.watchRecursively(notify = true)
                        } else {
                            logger.fine { "Received event ${event.kind().name()} for file $changedFile" }
                            onChange(FileSystemEvent(changedFile, event.toEventType()))
                        }
                    }
                }

                // reset the key to watch for future events
                val keyIsValid = watchKey.reset()
                if (!keyIsValid) {
                    logger.fine { "$directory has been unregistered" }
                    if (directory == workspace) {
                        logger.info("Workspace has been unregistered. Will stop watch service")
                        // if the workspace directory is not valid anymore it was either deleted or the watchService was stopped
                        // in either case, this watcher is done with its work
                        active.set(false)
                    }
                }
            }
        }
    }

    fun stop() {
        logger.info("Stopping watch service on workspace $workspace")
        active.set(false)
        watchService.close()
    }
}

enum class EventType {
    CREATED,
    MODIFIED,
    DELETED,
}

data class FileSystemEvent(
    val path: Path,
    val eventType: EventType,
)

private fun WatchEvent<*>.toEventType() =
    when (kind()) {
        StandardWatchEventKinds.ENTRY_MODIFY -> EventType.MODIFIED
        StandardWatchEventKinds.ENTRY_DELETE -> EventType.DELETED
        StandardWatchEventKinds.ENTRY_CREATE -> EventType.CREATED
        else -> throw IllegalStateException("Unknown event kind ${kind()} received")
    }
